File: bj/2775.py
import sys

def tenants(level, room):
    if records[level][room] > 0:
        return records[level][room]
    sum = 0
    for i in range(1, room+1):
        sum += tenants(level-1, i)
    records[level][room] = sum
    return sum

t = int(sys.stdin.readline().strip())

for i in range(t):
    k = int(sys.stdin.readline().strip())
    n = int(sys.stdin.readline().strip())
    records = [[_ for _ in range(n+1)]] + [[0 for _ in range(n+1)] for _ in range(k)]
    print(tenants(k, n))

File: bj/2775.cpp
#include<iostream>
#include<vector>
using namespace std;

int tenants(int level, int room, vector<vector<int>>& matrix){
    if (matrix[level][room] > 0) {
        return matrix[level][room];
    }
    int sum = 0;
    for (int i=1; i < room+1; i++){
        sum += tenants(level-1, i, matrix);
    }
    matrix[level][room] = sum;
    return sum;
}

int main(){
    int t;
    cin >> t;
    for (int i = 0; i < t; i++){
        int k, n;
        cin >> k;
        cin >> n;
        vector<vector<int>> records(k+1, vector<int>(n+1, 0));
        for (int j = 1; j < n+1; j++){
            records[0][j] = j;
        }
        cout << tenants(k, n, records) << endl;
    }
    return 0;
}